torch/nn/modules/module.py

```diff
@@ -193,13 +193,38 @@ def _apply(self, fn):
         for module in self.children():
             module._apply(fn)
 
-        for param in self._parameters.values():
+        def compute_should_use_data(tensor, tensor_applied):
+            if torch._has_same_tensorimpl_type(tensor, tensor_applied):
+                # If the new tensor has the same TensorImpl type as the existing tensor,
+                # then we take `torch.__future__.get_overwrite_module_params_on_conversion()`
+                # into account.
+                return not torch.__future__.get_overwrite_module_params_on_conversion()
+            else:
+                # If the new tensor has a different TensorImpl type, we always overwrite
+                # the existing tensor.
+                return False
+
+        for key, param in self._parameters.items():
             if param is not None:
-                # Tensors stored in modules are graph leaves, and we don't
-                # want to create copy nodes, so we have to unpack the data.
-                param.data = fn(param.data)
-                if param._grad is not None:
-                    param._grad.data = fn(param._grad.data)
+                # Tensors stored in modules are graph leaves, and we don't want to
+                # track autograd history of `param_applied`, so we have to use
+                # `with torch.no_grad():`
+                with torch.no_grad():
+                    param_applied = fn(param)
+                should_use_data = compute_should_use_data(param, param_applied)
+                if should_use_data:
+                    param.data = param_applied
+                else:
+                    assert isinstance(self._parameters[key], Parameter)
+                    self._parameters[key] = Parameter(param_applied, param.requires_grad)
+
+                if param.grad is not None:
+                    with torch.no_grad():
+                        grad_applied = fn(param.grad)
+                    if should_use_data:
```

Comment:
is there a reason we don't recompute should_use_data here?  If so, comment?

Reply:
I think recomputing `should_use_data` is a better idea, because we currently allow dense tensors to have sparse gradients. I changed the code accordingly to reflect this.

Reply:
The use case we are interested in is the following:

A parameter of the module is a dense tensor, but it has sparse gradient. If we use `._apply()` to convert this parameter and its gradient to a dense type, for the parameter we can use `set_data`, but for the gradient we have to overwrite the original tensor. Similarly, if we use `._apply()` to convert this parameter and its gradient to a sparse type, for the parameter we have to overwrite the original tensor, but for the gradient we can use `set_data`.

I think it's okay to allow the above behavior, because currently type/device conversion for modules with dense parameters and sparse gradients isn't working anyway (since `.data =` has to be called on the same TensorImpl type). The way we deal with sparse-dense conversion is also the same as how we do it if the `torch.__future__.get_overwrite_module_params_on_conversion()` flag is true. So I think it's okay to recompute `should_use_data` here.